Mathlib/GroupTheory/GroupAction/Blocks.lean

```diff
@@ -381,6 +395,163 @@ theorem IsBlockSystem.of_normal {N : Subgroup G} [N.Normal] :
 
 end Normal
 
+section Fintype
+
+theorem Setoid.nat_sum {α : Type _} [Finite α] {c : Set (Set α)} (hc : Setoid.IsPartition c) :
+    (finsum fun x : c => Set.ncard (x : Set α)) = Nat.card α := by
+  classical
+  have := Fintype.ofFinite α
+  simp only [finsum_eq_sum_of_fintype, Nat.card_eq_fintype_card, ← Set.Nat.card_coe_set_eq]
+  rw [← Fintype.card_sigma]
+  refine' Fintype.card_congr (Equiv.ofBijective (fun x => x.snd : (Σ a : ↥c, a) → α) _)
+  constructor
+  -- injectivity
+  rintro ⟨⟨x, hx⟩, ⟨a, ha : a ∈ x⟩⟩ ⟨⟨y, hy⟩, ⟨b, hb : b ∈ y⟩⟩ hab
+  dsimp at hab
+  rw [hab] at ha
+  rw [Sigma.subtype_ext_iff]
+  simp only [Subtype.mk_eq_mk, Subtype.coe_mk]
+  apply And.intro _ hab
+  exact ExistsUnique.unique (hc.2 b) ⟨hx, ha⟩ ⟨hy, hb⟩
+  -- surjectivity
+  intro a
+  obtain ⟨x, ⟨hx, ha : a ∈ x⟩, _⟩ := hc.2 a
+  use ⟨⟨x, hx⟩, ⟨a, ha⟩⟩
+
+theorem Set.ncard_coe {α : Type*} (s : Set α) :
+    s.ncard = Set.ncard (Set.univ : Set (Set.Elem s)) := by
+  apply Set.ncard_congr (fun a ha ↦ ⟨a, ha⟩)
+  · exact fun a ha ↦ by simp only [Set.mem_univ]
+  · simp [Subtype.mk_eq_mk]
+  · exact fun ⟨a, ha⟩ _ ↦ ⟨a, ha, rfl⟩
+
+/-- The cardinality of the ambient is the product of
+  of the cardinality of a block
+  by the cardinality of the set of translates of that block -/
+theorem IsBlock.ncard_block_mul_ncard_orbit_eq
+    [Finite X] [IsPretransitive G X] {B : Set X}
```

Comment:
It seems that many of the following lemmas have these assumptions. Would it make sense to start a
```
namespace IsBlock

variable [Finite X] [IsPretransitive G X] {B : Set X}
```
and then also remove the `IsBlock` from the names?

Reply:
I did that.

Reply:
Please resolve conversations when you've made the changes, to make it easier for later reviewers to see the state of the PR.

Reply:
OK. I was waiting for Johann's approval.